Add docs page for smoke tests (#2057)
smithy-lang/smithy: docs/source-2.0/additional-specs/smoke-tests.rst

.. _smoke-tests:

===========
Smoke Tests
===========

Smoke tests are small, simple tests intended to uncover large issues by ensuring
basic functionality works as expected. In Smithy, these tests are used to make
sure clients can successfully make a request to a live service, and that the

> service responds as expected.

Reviewer: nit: there are some phrasing inconsistencies in this PR for "service responds as expected", "service should respond to a given request", "get the right kind of response"

--------
Overview
--------

> This specification defines a single trait in the ``smithy.test`` namespace that is
> used to make basic assertions about how the service should respond to a given
> request from the client.

Reviewer: nit: omit "basic" and use "expectations" since that is what the trait uses

```suggestion
This specification defines a single trait in the ``smithy.test`` namespace that is
used to make expectations about how a service should respond to a given
request from a client.
```

:ref:`smithy.test#smokeTests <smokeTests-trait>`

>     Defines a set of test cases to send to a live service to ensure that a

Reviewer: nit: is `live` needed here?

Author: Maybe? This contrasts with, for example, protocol tests which don't run against live services. Maybe not entirely necessary but I don't think it hurts to be more specific here

    client can successfully connect to a service and get the right kind of

>     response.

Author: Expected response is a bit misleading - smoke tests don't make assertions about the actual content of the response, but what whether its successful or not, or whether the response is of the expected error type

This trait can be used by client and service developers to make sure their clients
can successfully connect to the service, and the service responds as expected. The

> trait can be used by client code generators to generate test cases that are run on
> the generated client against a service to ensure that this basic functionality is

Reviewer: nit:

```suggestion
trait can be used by client code generators to generate test cases that test
generated clients against services to ensure that critical functionality is
```

Author: I completely reworded this section, let me know what you think

working, and continues to work as the client and service evolve.

.. smithy-trait:: smithy.test#smokeTests
.. _smokeTests-trait:

----------
smokeTests
----------

Summary
    The ``smokeTests`` trait is used to define a set of test cases to send
    to a live service to ensure that a client can successfully connect to

>     a service and get the right kind of response.

Author: Same as https://github.com/smithy-lang/smithy/pull/2057#discussion_r1412178903, but I did update the wording a bit here

    * - vendorParams
      - ``document``
      - Defines vendor-specific parameters that are used to influence the
        request. For example, some vendors might utilize environment variables,
        configuration files on disk, or other means to influence the
        serialization formats used by clients or servers.

        If a ``vendorParamsShape`` is set, these parameters MUST be compatible
        with that shape's definition.
    * - vendorParamsShape
      - ``string``
      - A shape to be used to validate the ``vendorParams`` member contents.

        If set, the parameters in ``vendorParams`` MUST be compatible with this
        shape's definition.
    * - expect
      - :ref:`Expectation <Expectation-union>`
      - **Required**. Defines the response that is expected from the service
        call. This can be either a successful response, an error message, or
        a specific error response.
    * - tags
      - ``[string]``
      - Attaches a list of tags that can be used to categorize and group
        test cases. If a test case uses a feature that requires special
        configuration, it should be tagged.

.. _Expectation-union:

Expectation
-----------

> A union describing the different kinds of expectations that can be made for a
> test case. As it is a union, exactly one member must be set.

Reviewer: nit:

```suggestion
A union describing the expectations that can be made for a
test case. Exactly one member must be set.
```

.. list-table::
    :header-rows: 1
    :widths: 10 25 65

    * - Property
      - Type
      - Description
    * - success
      - ``Unit``
      - Indicates that the call is expected to not throw an error. No other

>         assertions are made about the response.

Author: I prefer "assertion" in the context of what the test case is doing - ie. the test case asserts x, rather than the test case expects x.

    * - failure
      - :ref:`FailureExpectation <FailureExpectation-struct>`

>       - Indicates that the call is expected to throw an error.

Reviewer: nit: not in the scope of this PR, but do we want to mark `Expectation` as `@unstable`?
I feel like there could be different types of expectations we could add in the future.

.. _FailureExpectation-struct:

FailureExpectation
~~~~~~~~~~~~~~~~~~

.. list-table::
    :header-rows: 1
    :widths: 10 25 65

    * - Property
      - Type
      - Description
    * - errorId
      - ``string``
      - Indicates that the call is expected to throw a specific type of error
        matching the targeted shape. If not specified, the error can be of
        any type.

Smoke tests example
===================

The following example defines smoke test cases for an operation which should
return a specific ``InvalidMessageError`` response when given an invalid input,
and return a successful response otherwise.

.. code:: smithy

    $version: "2"

    namespace smithy.example

    use smithy.test#smokeTests

    @smokeTests(
        [
            {
                id: "FooSuccess"

>                 params: {bar: "2"}

Reviewer: nit: not in the scope of the PR, but is there no `description` member for smoke tests? The ID by itself is not very descriptive.

                expect: {
                    success: {}
                }
            }
            {
                id: "FooInvalidMessageError"

>                 params: {bar: "föö"}

Reviewer: nit: here would be a good place to have a `description` member, maybe saying why the message is considered invalid

                expect: {
                    failure: {errorId: InvalidMessageError}
                }
            }
        ]
    )
    operation Foo {
        input: FooInput
        errors: [
            InvalidMessageError
        ]
    }

    @input
    structure FooInput {
        bar: String

>     }

Reviewer: nit: use inline I/O?
